### PR TITLE
fix deprecated array constructors

### PR DESCRIPTION
**Pull request by fredrikekre**

### CODE REVIEW

**Comment by fredrikekre** [comment]
Waiting for https://github.com/JuliaLang/METADATA.jl/pull/12295

**Comment by fredrikekre** [comment]
This will fail because of https://github.com/JuliaLang/Compat.jl/pull/416#issuecomment-348452332

**Comment by fredrikekre** [comment]
Well, tests pass on 0.6 now, but `KeyIterator` has changed to `KeySet` on master...

**Comment by oxinabox** [comment]
Needs to be rebased as of #349 which is a rough subset of these changes.
(My bad, I though that this was blocked waiting for #349, not that they directly overlapped)

**Comment by martinholters** [comment]
Ooops, yes, sorry, I had missed that you had updated this PR, too.

File: src/DataStructures.jl [hunk 1]
```diff
@@ -52,7 +59,11 @@ module DataStructures
 
     import Base: eachindex, keytype, valtype
 
-    _include_string(str) = eval(parse(str))
+    @static if VERSION < v"0.7.0-DEV.2437"
+        _include_string(str) = eval(parse(str))
+    else
+        _include_string(str) = eval(Meta.parse(str))
```
Comment (martinholters) [inline src/DataStructures.jl]: I think this can be used unconditionally, because `Meta.parse` is available on 0.6, too.

Comment (fredrikekre) [reply src/DataStructures.jl]: Ah, right.

**Comment by fredrikekre** [comment]
Rebased and removed the `KeySet` stuff from this PR.

File: src/ordered_dict.jl [hunk 2]
```diff
@@ -56,8 +56,8 @@ copy(d::OrderedDict) = OrderedDict(d)
 # OrderedDict{K,V}(kv::Tuple{Vararg{Tuple{K,V}}})          = OrderedDict{K,V}(kv)
 # OrderedDict{K  }(kv::Tuple{Vararg{Tuple{K,Any}}})        = OrderedDict{K,Any}(kv)
 # OrderedDict{V  }(kv::Tuple{Vararg{Tuple{Any,V}}})        = OrderedDict{Any,V}(kv)
-_oldOrderedDict1 = "OrderedDict{V}(kv::Tuple{Vararg{Pair{TypeVar(:K),V}}}) = OrderedDict{Any,V}(kv)"
-_newOrderedDict1 = "OrderedDict{V}(kv::Tuple{Vararg{Pair{K,V} where K}})   = OrderedDict{Any,V}(kv)"
+_oldOrderedDict1 = "OrderedDict(kv::Tuple{Vararg{Pair{TypeVar(:K),V}}}) where {V} = OrderedDict{Any,V}(kv)"
+_newOrderedDict1 = "OrderedDict(kv::Tuple{Vararg{Pair{K,V} where K}}) where {V}   = OrderedDict{Any,V}(kv)"
 OrderedDict(kv::Tuple{Vararg{Pair{K,V}}}) where {K,V}         = OrderedDict{K,V}(kv)
 OrderedDict(kv::Tuple{Vararg{Pair{K}}}) where {K}             = OrderedDict{K,Any}(kv)
 VERSION < v"0.6.0-dev.2123" ? _include_string(_oldOrderedDict1) : _include_string(_newOrderedDict1)
```
Comment (oxinabox) [inline src/ordered_dict.jl]: Julia <0.6 release is not supported, so this stuff can go, and just the real code can be put in.

File: src/ordered_dict.jl [hunk 3]
```diff
@@ -67,8 +67,8 @@ OrderedDict(kv::AbstractArray{Tuple{K,V}}) where {K,V} = OrderedDict{K,V}(kv)
 OrderedDict(kv::AbstractArray{Pair{K,V}}) where {K,V}  = OrderedDict{K,V}(kv)
 OrderedDict(kv::Associative{K,V}) where {K,V}          = OrderedDict{K,V}(kv)
 
-_oldOrderedDict2 = "OrderedDict{V}(ps::Pair{TypeVar(:K),V}...,) = OrderedDict{Any,V}(ps)"
-_newOrderedDict2 = "OrderedDict{V}(ps::(Pair{K,V} where K)...,) = OrderedDict{Any,V}(ps)"
+_oldOrderedDict2 = "OrderedDict(ps::Pair{TypeVar(:K),V}...,) where {V} = OrderedDict{Any,V}(ps)"
+_newOrderedDict2 = "OrderedDict(ps::(Pair{K,V} where K)...,) where {V} = OrderedDict{Any,V}(ps)"
 OrderedDict(ps::Pair{K,V}...) where {K,V}          = OrderedDict{K,V}(ps)
 OrderedDict(ps::Pair{K}...,) where {K}             = OrderedDict{K,Any}(ps)
 VERSION < v"0.6.0-dev.2123" ? _include_string(_oldOrderedDict2) : _include_string(_newOrderedDict2)
```
Comment (oxinabox) [inline src/ordered_dict.jl]: same

File: test/test_ordered_set.jl [hunk 4]
```diff
@@ -136,9 +136,9 @@
 
     # find
     s = OrderedSet([1,3,5,7])
-    @test findfirst(s,1) == 1
-    @test findfirst(s,7) == 4
-    @test findfirst(s,2) == 0
+    @test findfirst(equalto(1), s) == 1
+    @test findfirst(equalto(7), s) == 4
+    @test findfirst(equalto(2), s) == 0
```
Comment (oxinabox) [inline test/test_ordered_set.jl]: I guess these are coming in from `Compat` ?
It is a nice feature

Comment (fredrikekre) [reply test/test_ordered_set.jl]: Yes, https://github.com/JuliaLang/julia/pull/23812 on master and https://github.com/JuliaLang/Compat.jl/pull/405 in Compat.

**Comment by fredrikekre** [comment]
Removed the old code, which lets us remove the `_include_string` definition as well.

**Pull request merged**